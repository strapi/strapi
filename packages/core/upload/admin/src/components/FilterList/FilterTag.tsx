import { Tag } from '@strapi/design-system';
import { Cross } from '@strapi/icons';
import { useIntl } from 'react-intl';

import type { FilterStructure } from './FilterList';

type FilterTagAttribute = {
  fieldSchema?: {
    type?: string; // "date" | "enumeration"
    options?: {
      label: string; // "audio" | "video" | "image" | "file"
      value: string; // "audio" | "video" | "image" | "file"
    }[];
  };
  metadatas?: {
    label?: string; // "createdAt" | "updatedAt" | "type"
  };
  name: string; // "createdAt" | "updatedAt" | "mime"
};

interface FilterTagProps {
  attribute: FilterTagAttribute;
  operator: string;
  value: string;
  filter: FilterStructure;
  onClick: (filter: FilterStructure) => void;
}

export const FilterTag = ({ attribute, filter, onClick, operator, value }: FilterTagProps) => {
  const { formatMessage, formatDate, formatTime } = useIntl();

  const handleClick = () => {
    onClick(filter);
  };

  const { fieldSchema } = attribute;

  const type = fieldSchema?.type;

  let formattedValue = value;

  if (type === 'date') {
    formattedValue = formatDate(value, { dateStyle: 'full' });
  }

  if (type === 'datetime') {
    formattedValue = formatDate(value, { dateStyle: 'full', timeStyle: 'short' });
  }

  if (type === 'time') {
    const [hour, minute] = value.split(':');
    const date = new Date();
    date.setHours(Number(hour));
    date.setMinutes(Number(minute));

    formattedValue = formatTime(date, {
      hour: 'numeric',
      minute: 'numeric',
    });
  }

  const content = `${attribute.metadatas?.label} ${formatMessage({
    id: `components.FilterOptions.FILTER_TYPES.${operator}`,
    defaultMessage: operator,
  })} ${formattedValue}`;

  return (
    <Tag onClick={handleClick} icon={<Cross />} padding={1}>
      {content}
    </Tag>
  );
};
