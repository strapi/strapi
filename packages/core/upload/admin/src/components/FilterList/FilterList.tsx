// TODO: find a better naming convention for the file that was an index file before
/**
 *
 * FilterList
 *
 */
import { FilterTag } from './FilterTag';

type NumberKeyedObject = Record<number, string>;

type StringFilter = {
  [key: string]: string;
};

type MimeFilter = {
  [key: string]:
    | string
    | NumberKeyedObject
    | Record<string, string | NumberKeyedObject>
    | undefined;
};

export type FilterStructure = {
  [key: string]: MimeFilter | StringFilter | undefined;
};

export interface FilterListProps {
  appliedFilters: FilterStructure[];
  filtersSchema: {
    name: string;
    metadatas?: {
      label?: string;
    };
    fieldSchema?: {
      type?: string;
      mainField?: {
        name: string;
        type: string;
      };
      options?: {
        label: string;
        value: string;
      }[];
    };
  }[];
  onRemoveFilter: (filters: FilterStructure[]) => void;
}

export const FilterList = ({ appliedFilters, filtersSchema, onRemoveFilter }: FilterListProps) => {
  const handleClick = (filter: FilterStructure) => {
    const nextFilters = appliedFilters.filter((prevFilter) => {
      const name = Object.keys(filter)[0];
      const filterName = filter[name];
      if (filterName !== undefined) {
        const filterType = Object.keys(filterName)[0];
        const filterValue = filterName[filterType];
        if (typeof filterValue === 'string') {
          const decodedValue = decodeURIComponent(filterValue);
          return prevFilter[name]?.[filterType] !== decodedValue;
        }
      }

      return true;
    });

    onRemoveFilter(nextFilters);
  };

  return appliedFilters.map((filter, i) => {
    const attributeName = Object.keys(filter)[0];
    const attribute = filtersSchema.find(({ name }) => name === attributeName);

    if (!attribute) {
      // Handle the case where attribute is undefined
      return null;
    }

    const filterObj = filter[attributeName];
    const operator = Object.keys(filterObj!)[0];
    let value = filterObj![operator];

    if (Array.isArray(value)) {
      value = value.join(', ');
    } else if (typeof value === 'object') {
      value = Object.values(value).join(', ');
    } else {
      value =
        Array.isArray(value) || typeof value === 'object'
          ? Object.values(value).join(', ')
          : decodeURIComponent(value!);
    }

    let displayedOperator = operator;

    if (attribute?.name === 'mime') {
      displayedOperator = operator === '$contains' ? '$eq' : '$ne';

      // Type is file
      // The filter for the file is the following: { mime: {$not: {$contains: ['image', 'video']}}}
      if (operator === '$not') {
        value = 'file';
        displayedOperator = '$eq';
      }

      // Here the type is file and the filter is not file
      // { mime: {$contains: ['image', 'video'] }}
      if (['image', 'video'].includes(value[0]) && ['image', 'video'].includes(value[1])) {
        value = 'file';
        displayedOperator = '$ne';
      }
    }

    return (
      <FilterTag
        // eslint-disable-next-line react/no-array-index-key
        key={`${attributeName}-${i}`}
        attribute={attribute}
        filter={filter}
        onClick={handleClick}
        operator={displayedOperator}
        value={value as string}
      />
    );
  });
};
