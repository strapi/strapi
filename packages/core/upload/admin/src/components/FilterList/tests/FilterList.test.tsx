// TODO: find a better naming convention for the file that was an index file before
/**
 *
 * Tests for FilterList
 *
 */
import { DesignSystemProvider } from '@strapi/design-system';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import { FilterList } from '../FilterList';

const messages = {
  'components.FilterOptions.FILTER_TYPES.$eq': 'is',
  'components.FilterOptions.FILTER_TYPES.$ne': 'is not',
  'components.FilterOptions.FILTER_TYPES.$contains': 'contains (case sensitive)',
  'components.FilterOptions.FILTER_TYPES.$notContains': 'does not contain (case sensitive)',
  'components.FilterOptions.FILTER_TYPES.$gt': 'is greater than',
  'components.FilterOptions.FILTER_TYPES.$gte': 'is greater than or equal to',
  'components.FilterOptions.FILTER_TYPES.$lt': 'is lower than',
  'components.FilterOptions.FILTER_TYPES.$lte': 'is lower than or equal to',
  'components.FilterOptions.FILTER_TYPES.$startsWith': 'starts with',
  'components.FilterOptions.FILTER_TYPES.$endsWith': 'ends with',
  'components.FilterOptions.FILTER_TYPES.$null': 'is null',
  'components.FilterOptions.FILTER_TYPES.$notNull': 'is not null',
};

describe('<FilterList />', () => {
  it('renders and matches the snapshot', () => {
    const filters = [
      { mime: { $contains: 'image' } },
      { mime: { $notContains: 'image' } },
      { createdAt: { $eq: '2021-10-07' } },
      { mime: { $contains: 'video' } },
      {
        mime: {
          $not: {
            $contains: {
              0: 'image',
              1: 'video',
            },
          },
        },
      },
      {
        mime: {
          $contains: {
            0: 'image',
            1: 'video',
          },
        },
      },
    ];

    const { container } = render(
      <DesignSystemProvider>
        <IntlProvider locale="en" messages={messages} defaultLocale="en">
          <FilterList
            appliedFilters={filters}
            filtersSchema={[
              {
                name: 'createdAt',
                fieldSchema: {
                  type: 'date',
                },
                metadatas: { label: 'createdAt' },
              },
              {
                name: 'updatedAt',
                fieldSchema: {
                  type: 'date',
                },
                metadatas: { label: 'updatedAt' },
              },
              {
                name: 'mime',
                fieldSchema: {
                  type: 'enumeration',
                  options: [
                    { label: 'image', value: 'image' },
                    { label: 'video', value: 'video' },
                    { label: 'file', value: 'file' },
                  ],
                },
                metadatas: { label: 'type' },
              },
            ]}
            onRemoveFilter={jest.fn()}
          />
        </IntlProvider>
      </DesignSystemProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
