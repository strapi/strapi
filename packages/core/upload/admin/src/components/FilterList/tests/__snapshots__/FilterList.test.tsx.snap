// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterList /> renders and matches the snapshot 1`] = `
.c0 {
  padding-block-start: 4px;
  padding-inline-end: 4px;
  padding-block-end: 4px;
  padding-inline-start: 12px;
  height: 3.2rem;
  color: #4945ff;
  background: #f0f0ff;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  border-color: #d9d8ff;
}

.c4 {
  padding-block-start: 8px;
  padding-inline-end: 8px;
  padding-block-end: 8px;
  padding-inline-start: 8px;
}

.c1 {
  align-items: center;
  flex-direction: row;
  display: inline-flex;
}

.c2 {
  font-size: 1.2rem;
  line-height: 1.33;
  color: currentcolor;
  font-weight: 600;
}

.c6 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c5 {
  display: inline-flex;
  border: none;
}

.c5>svg {
  height: 1.2rem;
  width: 1.2rem;
}

.c5>svg path {
  fill: #4945ff;
}

.c5:hover {
  cursor: pointer;
}

.c3 {
  color: inherit;
  border-right: 1px solid #d9d8ff;
  padding-right: 8px;
}

<div>
  <div
    class="c0 c1"
  >
    <span
      class="c2 c3"
    >
      type is image
    </span>
    <button
      aria-disabled="false"
      class="c4 c5"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.061 23.939a1.503 1.503 0 0 1-2.125 2.125L16 18.125l-7.939 7.936a1.503 1.503 0 1 1-2.125-2.125L13.875 16 5.939 8.061a1.503 1.503 0 1 1 2.125-2.125L16 13.875l7.939-7.94a1.502 1.502 0 1 1 2.125 2.125L18.125 16z"
        />
      </svg>
    </button>
  </div>
  <div
    class="c0 c1"
  >
    <span
      class="c2 c3"
    >
      type is not image
    </span>
    <button
      aria-disabled="false"
      class="c4 c5"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.061 23.939a1.503 1.503 0 0 1-2.125 2.125L16 18.125l-7.939 7.936a1.503 1.503 0 1 1-2.125-2.125L13.875 16 5.939 8.061a1.503 1.503 0 1 1 2.125-2.125L16 13.875l7.939-7.94a1.502 1.502 0 1 1 2.125 2.125L18.125 16z"
        />
      </svg>
    </button>
  </div>
  <div
    class="c0 c1"
  >
    <span
      class="c2 c3"
    >
      createdAt is Thursday, October 7, 2021
    </span>
    <button
      aria-disabled="false"
      class="c4 c5"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.061 23.939a1.503 1.503 0 0 1-2.125 2.125L16 18.125l-7.939 7.936a1.503 1.503 0 1 1-2.125-2.125L13.875 16 5.939 8.061a1.503 1.503 0 1 1 2.125-2.125L16 13.875l7.939-7.94a1.502 1.502 0 1 1 2.125 2.125L18.125 16z"
        />
      </svg>
    </button>
  </div>
  <div
    class="c0 c1"
  >
    <span
      class="c2 c3"
    >
      type is video
    </span>
    <button
      aria-disabled="false"
      class="c4 c5"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.061 23.939a1.503 1.503 0 0 1-2.125 2.125L16 18.125l-7.939 7.936a1.503 1.503 0 1 1-2.125-2.125L13.875 16 5.939 8.061a1.503 1.503 0 1 1 2.125-2.125L16 13.875l7.939-7.94a1.502 1.502 0 1 1 2.125 2.125L18.125 16z"
        />
      </svg>
    </button>
  </div>
  <div
    class="c0 c1"
  >
    <span
      class="c2 c3"
    >
      type is file
    </span>
    <button
      aria-disabled="false"
      class="c4 c5"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.061 23.939a1.503 1.503 0 0 1-2.125 2.125L16 18.125l-7.939 7.936a1.503 1.503 0 1 1-2.125-2.125L13.875 16 5.939 8.061a1.503 1.503 0 1 1 2.125-2.125L16 13.875l7.939-7.94a1.502 1.502 0 1 1 2.125 2.125L18.125 16z"
        />
      </svg>
    </button>
  </div>
  <div
    class="c0 c1"
  >
    <span
      class="c2 c3"
    >
      type is image, video
    </span>
    <button
      aria-disabled="false"
      class="c4 c5"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.061 23.939a1.503 1.503 0 0 1-2.125 2.125L16 18.125l-7.939 7.936a1.503 1.503 0 1 1-2.125-2.125L13.875 16 5.939 8.061a1.503 1.503 0 1 1 2.125-2.125L16 13.875l7.939-7.94a1.502 1.502 0 1 1 2.125 2.125L18.125 16z"
        />
      </svg>
    </button>
  </div>
  <span
    class="c6"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;
